src/main/kotlin/com/exactpro/th2/check1/rule/AbstractCheckTask.kt

```diff
@@ -506,6 +520,18 @@ abstract class AbstractCheckTask(
         return checkpointTimeout == null || checkpointTimeout!!.isAfter(timestamp) || checkpointTimeout == timestamp
     }
 
+    /**
+     * Provides the ability to stop observing if a message timeout is set.
+     */
+    private fun Observable<Message>.onMessageTimeout() : Observable<Message> =
+        takeWhile {
+            checkOnMessageTimeout(it.metadata.timestamp).also { continueObservation ->
+                if (!continueObservation) {
+                    end(State.MESSAGE_TIMEOUT, "Expired message timeout")
+                }
```

Comment:
I believe we should not call the `end` method right here. We should do that only in terminated methods: onError, onComplete, the timeout is exited in timeout subscriber.

We can set a volatile variable with the status that should be passed when `onComplete` method will be called. By default, that status will be STREAM_COMPLETED but if this method is triggered the status will be changed to MESSAGE_TIMEOUT